File: server/src/_luaScriptsV2/syncDirty/markSyncDirty.lua
-- Merge selector/snapshot fields into the customer's sync dirty hash, refresh
-- its TTL, then attempt to set the signal marker (NX + TTL). Returns 1 when the
-- marker was newly set (no signal believed in flight), else 0.
-- uw:* fields carry {ts, snapshot} JSON: an older snapshot never overwrites a
-- newer one (cross-process batches can arrive out of order).
-- KEYS[1] dirty hash, KEYS[2] signal marker
-- ARGV[1] signal TTL seconds, ARGV[2] dirty TTL seconds, ARGV[3..] field/value pairs
local dirtyKey = KEYS[1]
local signalKey = KEYS[2]
local signalTtl = tonumber(ARGV[1])
local dirtyTtl = tonumber(ARGV[2])
for i = 3, #ARGV, 2 do
  local field = ARGV[i]
  local value = ARGV[i + 1]
  if string.sub(field, 1, 3) == 'uw:' then
    local existing = redis.call('HGET', dirtyKey, field)
    if not existing or cjson.decode(value).ts >= cjson.decode(existing).ts then
      redis.call('HSET', dirtyKey, field, value)
    end
  else
    redis.call('HSET', dirtyKey, field, value)
  end
end
redis.call('EXPIRE', dirtyKey, dirtyTtl)
local set = redis.call('SET', signalKey, '1', 'NX', 'EX', signalTtl)
if set then return 1 end
return 0

File: server/src/internal/balances/utils/sync/dirtyState/markSyncDirty.ts
import type { Redis } from "ioredis";
import type { UsageWindowUpdate } from "../../types/usageWindowUpdate.js";
import { buildSyncDirtyKeys, type SyncDirtyScope } from "./syncDirtyKeys.js";

/** Leak backstop: dirty state a worker never drains expires on its own.
 *  Reconciliation (sync conflicts / cache rebuild) covers anything expired. */
const DIRTY_STATE_TTL_SECONDS = 24 * 60 * 60;

/**
 * Hash-field encoding of the dirty state (single hash per customer so the
 * claim can RENAME it atomically):
 *   ce:{cusEntId}              = "1"        (selector set)
 *   ro:{rolloverId}            = "1"        (selector set)
 *   fc:{featureId}:{cusEntId}  = "1"        (feature -> cusEntIds map)
 *   uw:{featureId}             = JSON       ({ts, snapshot}, newest wins)
 *   entityId                   = string     (last write wins)
 */
export const DIRTY_FIELD_PREFIXES = {
	cusEnt: "ce:",
	rollover: "ro:",
	featureCusEnt: "fc:",
	usageWindow: "uw:",
	entityId: "entityId",
} as const;

// Merge fields into the dirty hash, refresh its TTL, then attempt to set the
// signal marker (NX + TTL). shouldSignal is true when the marker was newly
// set — i.e. no signal is believed to be in flight. A single Lua eval keeps
// merge+decision atomic against concurrent marks and claims.
import MARK_SCRIPT from "@/_luaScriptsV2/syncDirty/markSyncDirty.lua";

export const markSyncDirty = async ({
	redis,
	scope,
	cusEntIds,
	rolloverIds,
	modifiedCusEntIdsByFeatureId,
	usageWindowUpdates,
	entityId,
	signalTtlSeconds,
}: {
	redis: Redis;
	scope: SyncDirtyScope;
	cusEntIds: string[];
	rolloverIds: string[];
	modifiedCusEntIdsByFeatureId: Record<string, string[]>;
	usageWindowUpdates: Array<{ ts: number; update: UsageWindowUpdate }>;
	entityId?: string;
	signalTtlSeconds: number;
}): Promise<{ shouldSignal: boolean }> => {
	const { dirtyKey, signalKey } = buildSyncDirtyKeys(scope);

	const fieldArgs: string[] = [];
	for (const cusEntId of cusEntIds) {
		fieldArgs.push(`${DIRTY_FIELD_PREFIXES.cusEnt}${cusEntId}`, "1");
	}
	for (const rolloverId of rolloverIds) {
		fieldArgs.push(`${DIRTY_FIELD_PREFIXES.rollover}${rolloverId}`, "1");
	}
	for (const [featureId, featureCusEntIds] of Object.entries(
		modifiedCusEntIdsByFeatureId,
	)) {
		for (const cusEntId of featureCusEntIds) {
			fieldArgs.push(
				`${DIRTY_FIELD_PREFIXES.featureCusEnt}${encodeURIComponent(featureId)}:${encodeURIComponent(cusEntId)}`,
				"1",
			);
		}
	}
	for (const { ts, update } of usageWindowUpdates) {
		fieldArgs.push(
			`${DIRTY_FIELD_PREFIXES.usageWindow}${encodeURIComponent(update.feature_id)}`,
			JSON.stringify({ ts, snapshot: update }),
		);
	}
	if (entityId) {
		fieldArgs.push(DIRTY_FIELD_PREFIXES.entityId, entityId);
	}

	if (fieldArgs.length === 0) return { shouldSignal: false };

	const result = (await redis.eval(
		MARK_SCRIPT,
		2,
		dirtyKey,
		signalKey,
		signalTtlSeconds.toString(),
		DIRTY_STATE_TTL_SECONDS.toString(),
		...fieldArgs,
	)) as number;

	return { shouldSignal: result === 1 };
};
